backend/api/src/realms.rs

```diff
@@ -25,73 +38,126 @@ impl Realm {
         &self.name
     }
 
-    fn parent_id(&self) -> Id {
-        Id::realm(self.parent_key)
+    fn parent_id(&self) -> Option<Id> {
+        self.parent_key.map(Id::realm)
+    }
+
+    fn path(&self, context: &Context) -> String {
+        let mut segments = self.walk_up(context)
+            .map(|realm| &*realm.path_segment)
+            .collect::<Vec<_>>();
+        segments.reverse();
+        segments.join("/")
     }
 
-    fn parent(&self, context: &Context) -> &Realm {
-        &context.realm_tree.realms[&self.parent_key]
+    fn parent(&self, context: &Context) -> Option<&Realm> {
+        self.parent_key.map(|parent_key| &context.realm_tree.realms[&parent_key])
+    }
+
+    fn parents(&self, context: &Context) -> Vec<&Realm> {
+        let mut parents = self.walk_up(context).skip(1).collect::<Vec<_>>();
+        parents.reverse();
+        parents
     }
 
     fn children(&self, context: &Context) -> Vec<&Realm> {
-        context.realm_tree.children.get(&self.key)
-            .map(|children| {
-                children.iter()
-                    .map(|child_key| &context.realm_tree.realms[&child_key])
-                    .collect()
-            })
-            .unwrap_or_default()
+        self.child_keys.iter()
+            .map(|child| &context.realm_tree.realms[&child])
+            .collect()
     }
 }
 
 pub(crate) struct Tree {
     pub(crate) realms: HashMap<u64, Realm>,
-    children: HashMap<u64, Vec<u64>>,
+    from_path: HashMap<String, u64>,
```

Comment:
Mhhh shouldn't `u64` be `Key`?

Reply:
Yessss, and in may other places as well, actually. :grimacing: